docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,421 @@
+
+## Considered Options
+
+## Decision Outcome
+
+## Solution
+As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
+### Use cases
+After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:
+
+1. Keeping traffic in single zone (disable all cross-zone traffic)
+2. Load balancing based on priorities within a zone. Example: node → AZ → region
+3. Prefer traffic in single zone with fallback to other zones
+4. Load balance equally to list of zones
+5. When leaving zone, target specific zones first. Example zones physically close to each other
+6. Load balance equally to local dpp and other zone
+7. Exclude zone from load balancing
+8. Load balance equally within a group of zones
+
+### API design
+Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
+Because of this we have decided that we will split our API in two sections: `localZone ` and `crossZone`.
+
+Here is the full example of policy:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      disabled: false                                       # DEPRECATED
+      localZone:
+        affinityTags: ["k8s.io/node", "k8s.io/az"]          # (1)
+      crossZone:                                            # (2)
+        failover:                                           # (3)
+          - from:                                           # (4)
+              zones: ["zone-1", "zone-2"]
+            to:                                             # (5)
+              type: Only                                    # (6)
+              zones: ["zone-1", "zone-2"]
+          - from:
+              zones: ["zone-3"]
+            to:
+              type: AnyExcept
+              zones: ["zone-1"]
+          - to:
+              type: Any
+      overprovisioningFactor: 
+        percentage: 120                                     # (7)
+```
+
+
+(1) In `localZone` section, you configure the affinity of tags used for load balancing. If list is empty or localZone is not specified, load balancing works normally, without taking any priorities into account. 
+
+(2) In `crossZone` section, you configure zones priority.
+
+(3) In `failover` section, you configure list of zone priorities, first rules will have the highest load balancing priority and last in order will have the lowest priority.
+
+(4) In `from` section, you configure to which zones this rule applies. This rule will apply to all DPPs in `zone-1` and `zone-2`
```

Comment:
If empty it means all zones or do we make a semantic difference between `empty` and `omitted` ?

Reply:
I've added the default. I think empty from doesn't make sense so we should make `from.zones: []` empty list. If empty object then all zones